crates/next-dev/benches/mod.rs

```diff
@@ -264,6 +227,63 @@ fn bench_hmr_internal(mut g: BenchmarkGroup<WallTime>, location: CodeLocation) {
     }
 }
 
+fn insert_code(contents: &mut String, code: &str, location: CodeLocation) -> Result<()> {
+    const PRAGMA_EFFECT: &str = "/* @turbopack-bench:effect */";
+    const PRAGMA_EVAL: &str = "/* @turbopack-bench:eval */";
+    match location {
+        CodeLocation::Effect => {
+            let a = contents
+                .find(PRAGMA_EFFECT)
+                .ok_or_else(|| anyhow!("unable to find effect pragma in {}", contents))?;
+            contents.insert_str(
+                a + PRAGMA_EFFECT.len(),
+                &format!("\nEFFECT = () => {{ {code} }};"),
+            );
+        }
+        CodeLocation::Evaluation => {
+            let a = contents
+                .find(PRAGMA_EVAL)
+                .ok_or_else(|| anyhow!("unable to find eval pragma in {}", contents))?;
+            contents.insert_str(a + PRAGMA_EVAL.len(), &format!("\n{code}"));
+        }
+    }
+
+    Ok(())
+}
+
+static CHANGE_TIMEOUT_MESSAGE: &str = "update was not registered by bundler";
+
+async fn make_change<'a, M>(
+    module: &Path,
+    guard: &mut PageGuard<'a>,
+    location: CodeLocation,
+    timeout_duration: Duration,
+    measurement: &M,
+) -> Result<M::Value>
+where
+    M: Measurement,
+{
+    let msg = format!("TURBOPACK_BENCH_CHANGE_{}", guard.app_mut().counter());
+    let code = format!("globalThis.{BINDING_NAME} && globalThis.{BINDING_NAME}('{msg}');");
+
+    // Keep the IO out of the measurement.
+    let mut contents = fs::read_to_string(module)?;
+    insert_code(&mut contents, &code, location)?;
+    fs::write(module, contents)?;
+
+    let start = measurement.start();
```

Comment:
I think that's incorrect, because as far as I know, the watcher handler code will be called sync within the write call. So inside of write you are actually executing app code.

```suggestion
    // Keep the IO out of the measurement.
    let mut contents = fs::read_to_string(module)?;
    insert_code(&mut contents, &code, location)?;

    let start = measurement.start();
    fs::write(module, contents)?;
```

Reply:
As far as I can tell, inotify (Linux) and kqueue (macOS) are pull-based, so I don't see how that would be the case for either of them. As for Windows, I'm not sure.

Reply:
hmm... I didn't find anything regarding that...